File: spiralMatrix.js
// Leetcode 54

/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function(matrix) {
  var result = [];
  var left = 0;
  var right = matrix[0].length - 1;
  var top = 0;
  var bottom = matrix.length - 1;
  while (left <= right && top <= bottom) {
    for (var i = left; i <= right; i++) {
      result.push(matrix[top][i]);
    }
    top++;
    for (var i = top; i <= bottom; i++) {
      result.push(matrix[i][right]);
    }
    right--;
    if (top <= bottom) {
      for (var i = right; i >= left; i--) {
        result.push(matrix[bottom][i]);
      }
      bottom--;
    }
    if (left <= right) {
      for (var i = bottom; i >= top; i--) {
        result.push(matrix[i][left]);
      }
      left++;
    }
  }
  return result;
};

File: spiralMatrix.py
# Leetcode 54

from typing import *

class Solution:
  def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
    result = []
    left = 0
    top = 0
    right = len(matrix[0]) - 1
    bottom = len(matrix) - 1

    while left <= right and top <= bottom:
      for i in range(left, right + 1):
        result.append(matrix[top][i])
      top += 1
      for i in range(top, bottom + 1):
        result.append(matrix[i][right])
      right -= 1
      if (top <= bottom):
        for i in range(right, left - 1, -1):
          result.append(matrix[bottom][i])
      bottom -= 1
      if (left <= right):
        for i in range(bottom, top - 1, -1):
          result.append(matrix[i][left])
      left += 1

    return result

s = Solution()
matrix = [[1,2,3,4],[5,6,7,8],[9,10,11,12]]
print(s.spiralOrder(matrix))